File: blasty-back/src/main/java/com/blasty/service/Interface/ReservationService.java
package com.blasty.service.Interface;

import com.blasty.dto.request.ReservationRequest;
import com.blasty.dto.response.ReservationResponse;
import com.blasty.model.Reservation;

import java.time.LocalDateTime;
import java.util.List;

public interface ReservationService {
    ReservationResponse createReservation(ReservationRequest request);
    ReservationResponse getReservationById(Long id);
    List<ReservationResponse> getAllReservations();
    ReservationResponse confirmReservation(Long id);
    ReservationResponse cancelReservation(Long id);
    void deleteReservation(Long id);
    List<ReservationResponse> getReservationsByClientId(Long clientId);
    ReservationResponse completeReservation(Long id);
}

File: blasty-frontend/src/app/core/services/reservation.service.ts
import { Injectable } from "@angular/core"
import { HttpClient } from "@angular/common/http"
import { Observable } from "rxjs"
import { ReservationRequest, ReservationResponse } from "../models/reservation.model"
import { environment } from "../../environment/environment"

@Injectable({
  providedIn: "root",
})
export class ReservationService {
  private apiUrl = `${environment.apiURL}`

  constructor(private http: HttpClient) {}

  // Create a new reservation
  createReservation(reservation: ReservationRequest): Observable<ReservationResponse> {
    return this.http.post<ReservationResponse>(`${this.apiUrl}/reservations`, reservation)
  }

  // Get all reservations (admin only)
  getAllReservations(): Observable<ReservationResponse[]> {
    return this.http.get<ReservationResponse[]>(`${this.apiUrl}/reservations`)
  }

  // Get a specific reservation by ID
  getReservationById(id: number): Observable<ReservationResponse> {
    return this.http.get<ReservationResponse>(`${this.apiUrl}/reservations/${id}`)
  }

  // Get all reservations for the current client
  getMyReservations(): Observable<ReservationResponse[]> {
    return this.http.get<ReservationResponse[]>(`${this.apiUrl}/reservations/my-reservations`)
  }

  // Cancel a reservation
  cancelReservation(id: number): Observable<ReservationResponse> {
    return this.http.put<ReservationResponse>(`${this.apiUrl}/reservations/${id}/cancel`, {})
  }

  // Confirm a reservation (admin only)
  confirmReservation(id: number): Observable<ReservationResponse> {
    return this.http.put<ReservationResponse>(`${this.apiUrl}/reservations/${id}/confirm`, {})
  }

  // Complete a reservation (admin only)
  completeReservation(id: number): Observable<ReservationResponse> {
    return this.http.put<ReservationResponse>(`${this.apiUrl}/reservations/${id}/complete`, {})
  }

  // Delete a reservation (admin only)
  deleteReservation(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/reservations/${id}`)
  }
}

